src/io/flutter/logging/FlutterLogView.java

```diff
@@ -344,6 +346,31 @@ private String countString(int count) {
     }
   }
 
+  private static class LogVerticalScrollChangeListener implements ChangeListener {
+    private volatile int oldScrollValue = 0;
+
+    @Override
+    public void stateChanged(ChangeEvent e) {
+      final BoundedRangeModel model = (BoundedRangeModel)e.getSource();
+      final int newScrollValue = model.getValue();
+      final boolean isScrollUp = newScrollValue < oldScrollValue;
+      final boolean isScrollToEnd = newScrollValue + model.getExtent() == model.getMaximum();
+      if (isScrollUp) {
+        onScrollUp();
+      }
+      else if (isScrollToEnd) {
+        onScrollToEnd();
+      }
+      oldScrollValue = newScrollValue;
+    }
+
+    protected void onScrollUp() {
```

Comment:
Maybe make this an abstract class and these methods abstract to make it clear that implementers should implement?  Alternatively, maybe a comment saying something like:

    // implemented in subclass

?

Reply:
Done, I make this class as an abstract class.